src/eos-updater-poll.c

```diff
@@ -153,65 +308,181 @@ metadata_fetch_finished (GObject *object,
+                        EosMetricsInfo *metrics)
+{
+  UpdateAndMetrics *uam = g_new0 (UpdateAndMetrics, 1);
+
+  if (update != NULL)
+    uam->update = g_object_ref (update);
+
+  if (metrics != NULL)
+    uam->metrics = g_object_ref (metrics);
+
+  return uam;
+}
+
+static void
+update_and_metrics_free (UpdateAndMetrics *uam)
+{
+  if (uam == NULL)
+    return;
+
+  g_clear_object (&uam->update);
+  g_clear_object (&uam->metrics);
+  g_free (uam);
+}
+
+static UpdateAndMetrics *
+get_latest_uam (EosUpdaterData *data,
+                GHashTable *source_to_uam,
+                gboolean with_updates)
+{
+  g_autoptr(GHashTable) latest = g_hash_table_new (NULL, NULL);
+  GHashTableIter iter;
+  gpointer name_ptr;
+  gpointer uam_ptr;
+  GDateTime *latest_timestamp = NULL;
+  gsize idx;
+
+  g_hash_table_iter_init (&iter, source_to_uam);
+  while (g_hash_table_iter_next (&iter, &name_ptr, &uam_ptr))
+    {
+      UpdateAndMetrics *uam = uam_ptr;
+      EosBranchFile *branch_file = uam->metrics->branch_file;
+
+      if (with_updates && uam->update == NULL)
+        continue;
+
+      if (latest_timestamp != NULL)
+        {
+          if (g_date_time_compare (branch_file->download_time,
+                                   latest_timestamp))
+            {
+              latest_timestamp = NULL;
+              g_hash_table_remove_all (latest);
+            }
+        }
+      if (latest_timestamp == NULL)
+        {
+          latest_timestamp = branch_file->download_time;
+          g_hash_table_insert (latest, name_ptr, uam_ptr);
+        }
+    }
+
+  for (idx = 0; idx < data->n_download_sources; ++idx)
+    {
+      const gchar *name = order_key_str[data->download_order[idx]];
+      UpdateAndMetrics *uam = g_hash_table_lookup (latest, name);
+
+      if (uam != NULL)
+        return uam;
+    }
+
+  return NULL;
+}
+
+static void
 metadata_fetch (GTask *task,
                 gpointer object,
                 gpointer task_data,
                 GCancellable *cancel)
 {
-  EosUpdater *updater = EOS_UPDATER (object);
   EosUpdaterData *data = task_data;
-  OstreeRepo *repo = data->repo;
-  OstreeRepoPullFlags flags = (OSTREE_REPO_PULL_FLAGS_COMMIT_ONLY);
   GError *error = NULL;
-  g_autofree gchar *remote = NULL;
-  g_autofree gchar *branch = NULL;
-  g_autofree gchar *refspec = NULL;
-  g_autofree gchar *orig_refspec = NULL;
-  gchar *pullrefs[] = { NULL, NULL };
-  gchar *csum = NULL;
   GMainContext *task_context = g_main_context_new ();
-  g_autoptr(GVariant) commit = NULL;
-
-  g_main_context_push_thread_default (task_context);
+  g_autoptr(EosMetadataFetchData) fetch_data = NULL;
+  g_autoptr(GPtrArray) fetchers = NULL;
+  guint idx;
+  UpdateAndMetrics *latest_uam = NULL;
+  g_autoptr(GHashTable) source_to_uam = NULL;
 
-  if (!eos_updater_get_upgrade_info (repo, &refspec, &orig_refspec, &error))
-    goto error;
+  fetch_data = eos_metadata_fetch_data_new (task, data, task_context);
+  g_main_context_unref (task_context);
```

Comment:
It is passed to `eos_metadata_fetch_data_new` (see the last parameter - `task_context`). This function just takes its own ref on the context and we don't need to keep another one anymore.